ocaml/xapi/repository.ml

```diff
@@ -333,92 +292,69 @@ let get_hosts_updates ~__context =
-        |> List.map Livepatch.of_json
-      in
-      (h, applied_livepatches)
-    )
-    hosts_updates
-
-(* Group updates by repository for each host *)
-let get_updates_of_hosts ~__context enabled_repositories hosts_updates =
-  List.map
-    (fun (h, updates_of_host) ->
-      group_host_updates_by_repository ~__context enabled_repositories h
-        updates_of_host
-    )
-    hosts_updates
+let get_applied_livepatches_of_host updates_of_host =
+  get_list_from_updates_of_host "applied_livepatches" updates_of_host
+  |> List.map Livepatch.of_json
 
-let is_livepatchable ~__context repository applied_livepatches_of_hosts =
+let is_livepatchable ~__context repository applied_livepatches_of_host =
   let updates_info =
     parse_updateinfo ~__context ~self:repository ~check:false
   in
   List.exists
-    (fun (_, applied_livepatches) ->
-      List.exists
-        (fun lp ->
-          get_accumulative_livepatches ~since:lp ~updates_info |> function
-          | [] ->
-              false
-          | _ ->
-              true
-        )
-        applied_livepatches
+    (fun lp ->
+      get_accumulative_livepatches ~since:lp ~updates_info |> function
+      | [] ->
+          false
+      | _ ->
+          true
     )
-    applied_livepatches_of_hosts
+    applied_livepatches_of_host
 
 let set_available_updates ~__context =
   ignore (get_single_enabled_update_repository ~__context) ;
   let enabled = get_enabled_repositories ~__context in
   let hosts_updates = get_hosts_updates ~__context in
-  let applied_livepatches_of_hosts =
-    get_applied_livepatches_of_hosts hosts_updates
-  in
-  let updates_of_hosts =
-    get_updates_of_hosts ~__context enabled hosts_updates
-  in
-  (* Group updates by repository for all hosts *)
-  let updates_by_repository =
-    List.fold_left
-      (fun acc l ->
-        List.fold_left (fun acc' (x, y) -> append_by_key acc' x y) acc l
-      )
-      [] updates_of_hosts
-  in
+  List.iter
+    (fun (h, updates_of_host) ->
+      let host_pkg_updates_available =
+        match get_list_from_updates_of_host "updates" updates_of_host with
+        | [] ->
+            false
+        | _ ->
+            true
+      in
+      let up_to_date =
+        match host_pkg_updates_available with
+        | true ->
+            false
+        | false ->
+            (* No RPM packages to be updated.
+             * Find out if there are available livepatches from a update repo
+             *)
+            let update_repo =
+              get_singleton
+                (List.filter
+                   (fun repo -> Db.Repository.get_update ~__context ~self:repo)
+                   enabled
+                )
+            in
+            not
+              (is_livepatchable ~__context update_repo
+                 (get_applied_livepatches_of_host updates_of_host)
+              )
+      in
+      let up_to_date_state =
+        match up_to_date with true -> `yes | false -> `no
```

Comment:
No need to match 3 times. Only one bool data is flowing through the whole logic. Would match be enough?